core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h

```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+
+  mp_buffer_info_t extra_commit;
+  mp_get_buffer_raise(args[6], &extra_commit, MP_BUFFER_READ);
+
+  mp_buffer_info_t gen;
+  mp_get_buffer_raise(args[7], &gen, MP_BUFFER_READ);
+  if (gen.len != 33) {
+    mp_raise_ValueError("Invalid length of generator");
+  }
+  secp256k1_generator generator;
+  if (!secp256k1_generator_parse(ctx, &generator, gen.buf)) {
+    mp_raise_ValueError("Generator parsing failed");
+  }
+
+  mp_buffer_info_t scratch_buffer;
+  mp_get_buffer_raise(args[8], &scratch_buffer, MP_BUFFER_WRITE);
+  if (scratch_buffer.len < RANGEPROOF_SIGN_BUFFER_SIZE) {
+    mp_raise_ValueError("Invalid length of output buffer");
+  }
+
+  size_t rangeproof_len = scratch_buffer.len;
+  if (!secp256k1_rangeproof_sign(
+          ctx, scratch_buffer.buf, &rangeproof_len, VALUE_MIN, &commitment,
+          blind.buf, nonce.buf, EXPONENT, BITS, value, message.buf, message.len,
+          extra_commit.buf, extra_commit.len, &generator)) {
+    mp_raise_ValueError("Rangeproof sign failed");
+  }
+  return mp_obj_new_memoryview('B', rangeproof_len, scratch_buffer.buf);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(
+    mod_trezorcrypto_secp256k1_context_rangeproof_sign_obj, 9, 9,
+    mod_trezorcrypto_secp256k1_context_rangeproof_sign);
+
+/// def rangeproof_rewind(self, conf_value: bytes, conf_asset: bytes,
+///                       nonce: bytes, range_proof: bytes,
+///                       extra_commit: bytes, asset_message: bytearray) ->
+///                       (value: long, blind: bytes,
+///                        asset_message: memoryview):
+///     '''
+///     Rewind a range proof to get the value, blinding factor and message.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_rangeproof_rewind(
+    size_t n_args, const mp_obj_t *args) {
+  const secp256k1_context *ctx =
+      mod_trezorcrypto_get_secp256k1_context(args[0]);
+
+  mp_buffer_info_t conf_value;
+  mp_get_buffer_raise(args[1], &conf_value, MP_BUFFER_READ);
+  secp256k1_pedersen_commitment commitment;
+  if (!secp256k1_pedersen_commitment_parse(ctx, &commitment, conf_value.buf)) {
+    mp_raise_ValueError("Invalid Pedersen commitment");
+  }
+
+  mp_buffer_info_t gen;
+  mp_get_buffer_raise(args[2], &gen, MP_BUFFER_READ);
+  if (gen.len != 33) {
+    mp_raise_ValueError("Invalid length of confidential asset");
+  }
+  secp256k1_generator generator;
+  if (!secp256k1_generator_parse(ctx, &generator, gen.buf)) {
+    mp_raise_ValueError("Generator parsing failed");
+  }
+
+  mp_buffer_info_t nonce;
+  mp_get_buffer_raise(args[3], &nonce, MP_BUFFER_READ);
+
+  mp_buffer_info_t range_proof;
+  mp_get_buffer_raise(args[4], &range_proof, MP_BUFFER_READ);
+
+  mp_buffer_info_t extra_commit;
+  mp_get_buffer_raise(args[5], &extra_commit, MP_BUFFER_READ);
+
+  mp_buffer_info_t asset_message;
+  mp_get_buffer_raise(args[6], &asset_message, MP_BUFFER_WRITE);
+
+  size_t asset_message_len = asset_message.len;
+  byte value_blind[32] = {0};
```

Comment:
(no need to init, also no need in the following three uint64_t declarations)

Reply:
Fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871.